velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,320 @@
+    const uint32_t index,
+    const RowColumn& rowColumn,
+    char* const row,
+    char* const prefix) {
+  normalizeNumericColumn<
+      TypeKind::INTEGER,
+      typename TypeTraits<TypeKind::INTEGER>::NativeType>(
+      prefixSortLayout, index, rowColumn, row, prefix);
+}
+} // namespace
+
+PrefixSortLayout PrefixSortLayout::makeSortLayout(
+    const std::vector<TypePtr>& types,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const uint32_t maxNormalizedKeySize) {
+  VELOX_CHECK_GE(types.size(), 0);
+  uint32_t normalizedKeySize = 0;
+  uint32_t numNormalizedKeys = 0;
+  uint32_t numSortKeys = types.size();
+  std::vector<uint32_t> prefixOffsets;
+  for (const auto& type : types) {
+    if (normalizedKeySize > maxNormalizedKeySize) {
+      break;
+    }
+    if (type->kind() == TypeKind::BIGINT) {
+      prefixOffsets.push_back(numNormalizedKeys++);
+      normalizedKeySize += sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+    } else if (type->kind() == TypeKind::INTEGER) {
+      prefixOffsets.push_back(numNormalizedKeys++);
+      normalizedKeySize += sizeof(TypeTraits<TypeKind::INTEGER>::NativeType);
+    } else {
+      break;
+    }
+  }
+  uint32_t compareMask = normalizedKeySize % 8;
+  uint32_t remainIntOffset = (normalizedKeySize / 8) * 8;
+  uint32_t remainShortOffset = remainIntOffset + (compareMask & 0b100) * 4;
+  uint32_t remainByteOffset = remainShortOffset + (compareMask & 0b010) * 2;
+
+  return PrefixSortLayout{
+      normalizedKeySize + sizeof(char*),
+      normalizedKeySize,
+      numNormalizedKeys,
+      numSortKeys,
+      keyCompareFlags,
+      numNormalizedKeys < numSortKeys,
+      std::move(prefixOffsets),
+      compareMask,
+      remainIntOffset,
+      remainShortOffset,
+      remainByteOffset};
+}
+
+// For normalizedKeySize > 8, use uint64_t word compare.
+// For remain part of the binary string, the size has only 7 cases, and
+// any number less than 8 can be represented by 3 bits:
+// highest bit means 4, we use integer compare,
+// middle bit means 2, we use memcmp(char*, char*, 2)
+// lowest bit means 1, we use memcmp(char*, char*, 1)
+FOLLY_ALWAYS_INLINE int
+PrefixSort::memCompareByWord(char* left, char* right, int size) {
+  for (int i = 0; i < size / 8; ++i) {
+    uint64_t v1 = *reinterpret_cast<uint64_t*>(left + i * 8);
+    uint64_t v2 = *reinterpret_cast<uint64_t*>(right + i * 8);
+    // Use compare to avoid data overflow.
+    if (v1 > v2) {
+      return 1;
+    } else if (v1 < v2) {
+      return -1;
+    }
+  }
+  if (sortLayout_.remainBytesToCompare & 0b100) {
+    uint32_t v1 =
+        *reinterpret_cast<uint32_t*>(left + sortLayout_.remainIntOffset);
+    uint32_t v2 =
+        *reinterpret_cast<uint32_t*>(right + sortLayout_.remainIntOffset);
+    // Use compare to avoid data overflow.
+    if (v1 > v2) {
+      return 1;
+    } else if (v1 < v2) {
+      return -1;
+    }
+  }
+  if (sortLayout_.remainBytesToCompare & 0b010) {
```

Comment:
Any reason not to use std::memcmp for all of the remaining bytes?

Reply:
Assuming that memcmp(char* , char*, size) is slower than memcmp(char* , char*, constant_size),  constant_size can be determined at the ’compilation time‘

Reply:
got it , I will open a new PR to modify it